tools/ir-generator/irclass.h

```diff
@@ -169,10 +178,24 @@ class IrField : public IrElement {
     IrField(const Type *type, cstring name, int flags)
         : IrField(Util::SourceInfo(), type, name, cstring(), flags) {}
     void resolve() override;
-    void generate(std::ostream &out, bool asField) const;
+    virtual void generate(std::ostream &out, bool asField) const;
     void generate_hdr(std::ostream &out) const override { generate(out, true); }
     void generate_impl(std::ostream &) const override;
     cstring toString() const override { return name; }
+
+ protected:
+    void resolveType(const Type *type);
+};
+
+class IrVariantField : public IrField {
+ public:
+    std::vector<const Type *> *types;
```

Comment:
I experimented with this, but this turned out to be awkward – variant is type, but we need to generate extra boilerplate for variant fields. So instead of duplication of things (type & field) I decided to keep it in the one place. And null type in `IrField` appeared to be very handy with implementations as for variants we just segfault :)